drop-in/src/main/java/com/adyen/checkout/dropin/ui/DropInViewModel.kt

```diff
@@ -10,84 +10,184 @@ package com.adyen.checkout.dropin.ui
+            throw CheckoutException("Concurrency error. Cannot update Payment methods list because availability is still being checked.")
+        }
+
+        // Reset variables
+        availabilitySum = 0
+        availabilitySkipSum = 0
+        availabilityChecksum = paymentMethods.size
+        paymentMethodsList.clear()
 
         for (paymentMethod in paymentMethods) {
             val type = paymentMethod.type
+            when {
+                type == null -> {
+                    throw CheckoutException("PaymentMethod type is null")
+                }
+                PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type) -> {
+                    Logger.v(TAG, "Supported payment method: $type")
+                    // We assume payment method is available and remove it later when the callback comes
+                    // this is the overwhelming majority of cases, and we keep the list ordered this way.
+                    paymentMethodsList.add(
+                        PaymentMethodModel(type, paymentMethod.name ?: nullNameFallback)
+                    )
+                    checkComponentAvailability(getApplication(), paymentMethod, dropInConfiguration, this)
+                }
+                else -> {
+                    availabilitySkipSum++
+                    if (PaymentMethodTypes.UNSUPPORTED_PAYMENT_METHODS.contains(type)) {
+                        Logger.e(TAG, "PaymentMethod not yet supported - $type")
+                    } else {
+                        Logger.d(TAG, "No details required - $type")
+                        paymentMethodsList.add(
+                            PaymentMethodModel(type, paymentMethod.name ?: nullNameFallback)
+                        )
+                    }
+                    // If last payment method is redirect list might be ready now
+                    checkIfListIsReady()
+                }
+            }
+        }
+    }
 
-            if (type == null) {
-                Logger.e(TAG, "PaymentMethod type is null")
-            } else if (PaymentMethodTypes.SUPPORTED_PAYMENT_METHODS.contains(type)) {
-                checkComponentAvailability(getApplication(), paymentMethod, dropInConfiguration, this)
-            } else {
-                if (!PaymentMethodTypes.UNSUPPORTED_PAYMENT_METHODS.contains(type)) {
-                    Logger.d(TAG, "No details required - $type")
-                    addPaymentMethod(paymentMethod)
-                } else {
-                    Logger.e(TAG, "PaymentMethod not yet supported - $type")
+    override fun onAvailabilityResult(isAvailable: Boolean, paymentMethod: PaymentMethod, config: Configuration?) {
+        Logger.d(TAG, "onAvailabilityResult - ${paymentMethod.type}: $isAvailable")
+
+        availabilitySum++
+
+        if (!isAvailable) {
+            Logger.e(TAG, "${paymentMethod.type} NOT AVAILABLE")
+            for ((index, paymentMethodModel) in paymentMethodsList.withIndex()) {
```

Comment:
We can do something like this: `paymentMethodsList.removeAll {  it.type == paymentMethod.type }` ?

Reply:
<3